include/pybind11/pybind11.h

```diff
@@ -900,30 +900,30 @@ class module : public object {
             py::module m2 = m.def_submodule("sub", "A submodule of 'example'");
             py::module m3 = m2.def_submodule("subsub", "A submodule of 'example.sub'");
     \endrst */
-    module def_submodule(const char *name, const char *doc = nullptr) {
+    module_ def_submodule(const char *name, const char *doc = nullptr) {
```

Comment:
This is the one that created the error. It is also a member function of `py::module`. This meant that going with `using module_ = module;` doesn't help and I had to rename the whole class.

Unless we were to go with `pybind11::module`, but @YannickJadoul seems to be against that.

Reply:
I think we should recommend `py::module`; this is just an internal change since `module` unqualified is not allowed in some cases. Users should not be exporting the pybind11 namespace. I don't think users should even know that a `module_` exists normally.

Reply:
> whichever way you choose to be the "real" class vs. the alias

He was fine with either way, just suggesting using `module_` internally instead of writing `pybind11::module` vs. unqualified names everywhere (internally only).

Reply:
Okay, I added a note in the changelog.

Reply:
> but @YannickJadoul seems to be against that.

I'm not against it, for the record. I probably prefer the other one, since it's more consistent with the rest of pybind11 :-)

> I think we should recommend py::module; this is just an internal change since module unqualified is not allowed in some cases.

Also true. So normally, users shouldn't even notice, and if they do have a `using pybind11::module;` or `using namespace pybind11;`, we can't really be blamed?
But explaining to use `module_` seems a lot easier than explaining when to use `py::module`? And it's more consistent with `class_` and `enum_` and `int_` and `float_`?

Reply:
`class`, `enum`, `int`, and `float` have been C++ keywords since the dawn of time (more or less). `module` is _slighly_ special in C++20, and is not a true keyword. And the restrictions on using it are the same for pybind11 as with any other code; including user code. Someone might already write `auto module = ...; module...`, and we can't fix/help that. I don't think we should try to force pretty much every pybind11 project to change the name of `py::module` unless we really had to (and `module` is easier to remember than `module_`). And I think 99% of pybind11 follows the examples and uses `py` as the namespace.

If C++ were to make this a true keyword in 3*n years (I doubt it), then we could force a transition. But I really doubt it, it's too common a term.

Reply:
No, I think I agree: no need to force anything. Just wondering if we should give the "safe" example and just start using `py::module_` in examples, etc; more from a perspective of nudging users than to tell them to transition.

If we don't want to do that, I'm also fine with `pybind11::module` in our code, btw. Thinking about it again, I don't really object to that approach either. Up to @bstaletic who took up the task of fixing this :-)

Reply:
From what I can tell, no one has a particularly strong opinion. The current patch should work, so why not go with it, as is and, if people start having troubles in their code, we can change the docs to use `module_` more consistently?

Any objections to that idea?

To be honest, I'm just try not to end up endlessly bikeshedding mostly equivalent alternatives.